Builders - adding a custom made module (#26)
moondance-labs/tanssi-docs: builders/build/substrate/adding-custom-made-module.md

Being the module an independent package, it has its own `Cargo.toml` where the module attributes and dependencies must be defined.

As an example of the attributes, the name of the module, version, authors, and other relevant information can be set.

```toml
[package]
name = "module-lottery-example"
version = "4.0.0-dev"
description = "Simple module example"
authors = [""]
homepage = ""
...
```

This file also defines the module's dependencies, such as the core functionality that allows seamless integration with the runtime and other modules, access to storage, events emission, and more. 

The full example of the `Cargo.toml` file sets, besides the attributes, the dependencies required by Substrate:

??? code "View the complete Cargo.toml file"

    ```rust
    --8<-- 'code/basic-substrate/lottery-example-cargo.toml'
    ```

## Adding Custom Logic {: #adding-custom-logic}

As presented in the [custom-made module](/learn/framework/modules/#custom-modules){target=_blank} section of the modularity article, creating a module involves implementing the following attribute macros, of which the first three are mandatory:

> 1. `#[frame_support::pallet]` - this attribute is the entry point that marks the module as usable in the runtime

Author: I tought about it while revisiting these lines, but as it is expressed differently in both articles (one is a numbered list, the other is a normal list and separated in two sections), the only matching text is the macro descriptions and it is easier to just have them copied than creating a snippet for each one. wdyt ?

    // All the logic goes here
}
```

### Implementing the Module Configuration {: #implementing-module-configuration }

To make the modules highly adaptable, their configuration is abstract enough to allow them to be adapted to the specific requirements of the use case the runtime implements.

The implementation of the `#[pallet::config]` macro is mandatory and sets the module's dependency on other modules and the types and values specified by the runtime-specific settings. More about module dependency in the [Substrate documentation](https://docs.substrate.io/build/pallet-coupling/){target=_blank}.

In this example, the lottery module depends on other modules to manage the currency and the random function to select the winner and also reads and uses the ticket price and the maximum number of participants directly from the runtime settings. The configurations for this module are:

1. Events: the module depends on the runtime's definition of an event to be able to emit them.
2. Currency: this module needs to transfer funds, hence, it needs the definition of the currency system from the runtime
3. Randomness: to fairly select the winner of the prize from the list of participants 

> 4. Ticket cost: the price to charge the buyers that participate in the lottery

Author: Substrate magic: in the runtime configuration we have a line "type TicketCost = ConstU128<1000000000000000>;" that defines the concrete type and value, so the pallet matches its abstract type definition with that concrete one defined in the runtime, using BalanceOf<> (instead of u128) to make it possible to arithmetically operate with the currency underlying platform (in this case, balances)

5. Maximum number of participants: the top limit of participants allowed in each lottery round
6. Module Id: the module unique identifier is required to access the module account to hold the participant's funds until transferred to the winner

The implementation of the described configuration for this example is shown in the following code snippet:

```rust
#[pallet::config]
pub trait Config: frame_system::Config {

    // 1. Event definition
    type RuntimeEvent: From<Event<Self>> 
        + IsType<<Self as frame_system::Config>::RuntimeEvent>;

>     // 2. Currency 

Author: We don't really need to include the lockable and reservable traits! (I deleted the references)

    type Currency: Currency<Self::AccountId> 
        + ReservableCurrency<Self::AccountId>
        + LockableCurrency<Self::AccountId>;

>     // 3. Randomness

Author: added some clarification on the internal mechanisms of the module. Randomness is another pallet, and this is the standar way os using it as a "loose" coupled dependency

    type MyRandomness: Randomness<Self::Hash, BlockNumberFor<Self>>;

    // 4. Ticket cost
    #[pallet::constant]
    type TicketCost: Get<BalanceOf<Self>>;

    // 5. Maximum number of participants
    #[pallet::constant]
    type MaxParticipants: Get<u32>;

    // 6. Module Id
    #[pallet::constant]
    type PalletId: Get<PalletId>;
}
```

This abstract definition of dependencies is crucial to avoid coupling to a specific use case, and to enable the modules to serve as basic building blocks to Substrate Appchains.

### Implementing Transactions {: #implementing-transactions } 

Calls represent the behavior a runtime exposes, in the form of transactions that can be dispatched for processing. By implementing the macro `#[pallet::call]`, the custom logic can be added to the module.

This is the general structure of the macro implementation, and the calls definition:

> ```rust

Author: Much better!

        // 4. Transfers the total prize to the winner's account
        // 5. Resets the participants list, and gets ready for another lottery round

    }
}
```

These calls also emit events, to keep the user informed and can return errors, should any of the validations go wrong.

Here is the complete implementation of the calls, with the custom lottery logic:

??? code "View the complete calls code"

    ```rust
    --8<-- 'code/basic-substrate/lottery-example-calls.rs'
    ```

### Implementing Custom Errors {: #implementing-custom-errors}

The `#[pallet::error]` macro is applied to an enumeration of errors that might occur during the execution. It is important for security reasons to handle all error cases gracefully and never crash in the runtime.

```rust
// Errors inform users that something went wrong.
#[pallet::error]
pub enum Error<T> {
    NotEnoughCurrency,
    AccountAlreadyParticipating,
    CanNotAddParticipant,
}
```

### Implementing Events {: #implementing-events }

The `#[pallet::event]` macro is applied to an enumeration of events to inform the user of any changes in the state or important actions that happened during the execution in the runtime.

> The following snippet shows the events defined in the lottery example:

Reviewer: ```suggestion
As an example, for the `lottery-example` module, this macro could be configured with the following events:
```

```rust
#[pallet::event]
#[pallet::generate_deposit(pub(super) fn deposit_event)]
pub enum Event<T: Config> {
    /// Event emitted when a ticket is bought
    TicketBought { who: T::AccountId },
    /// Event emitted when the prize is awarded
    PrizeAwarded { winner: T::AccountId },
    /// Event emitted when there are no participants
    ThereAreNoParticipants,
}
```

### Implementing Storage for State Persistence {: #implementing-storage }

> The `#[pallet::storage]` macro initializes a runtime storage structure. In this example, a basic value storage structure is used to persist the list of participants in a bounded capacity vector ([BoundedVec](https://crates.parity.io/frame_support/storage/bounded_vec/struct.BoundedVec.html){target=_blank}).

Reviewer: To follow the same structure as before I would add the next paragraph in this one so something like:

```suggestion
The `#[pallet::storage]` macro initializes a runtime storage structure.  In the heavily constrained environment of an AppChain, deciding what to store and which structure to use can be critical in terms of performance. More on this topic is covered in the [Substrate documentation](https://docs.substrate.io/build/runtime-storage/){target=_blank}.

In this example,  the `lottery-example` module needs a basic value storage structure used to persist the list of participants in a bounded capacity vector ([BoundedVec](https://crates.parity.io/frame_support/storage/bounded_vec/struct.BoundedVec.html){target=_blank}). This can be initialized as follows:
```

> In the heavily constrained environment of an AppChain, deciding what to store and which structure to use can be critical in terms of performance. More on this topic is covered in the [Substrate documentation](https://docs.substrate.io/build/runtime-storage/){target=_blank}.

Reviewer: Appchain or ContainerChain?

Author: Uhmm, It's a general statement, situated in a design phase I'd use Appchain

> In the heavily constrained environment of an AppChain, deciding what to store and which structure to use can be critical in terms of performance. More on this topic is covered in the [Substrate documentation](https://docs.substrate.io/build/runtime-storage/){target=_blank}.

Reviewer: If you agree with my suggestion above, then delete this paragraph

Author: Agreed!

```rust
#[pallet::storage]
#[pallet::getter(fn get_participants)]
pub(super) type Participants<T: Config> = StorageValue<
    _,
    BoundedVec<T::AccountId, T::MaxParticipants>,
    OptionQuery
>;
```
### The Complete Module {: #complete-module }

Putting all the pieces together, after implementing all the required macros and adding the custom logic, now the module is complete and ready to be used in the runtime. 

??? code "View the complete module file"

    ```rust
    --8<-- 'code/modules/lottery-example.rs'
    ```

## Configure the Runtime {: #configure-runtime }

> Finally, with the module finished, it can be included in the runtime. By doing so, the transactions buy_tickets and award_prize will be callable by the users.

Reviewer: ```suggestion
Finally, with the module finished, it can be included in the runtime. By doing so, the transactions `buy_tickets` and `award_prize` will be callable by the users. This also means that the (Polkadot.js API)(/builders/interact/substrate-api/polkadot-js-api/) will be decorated with this module module and all the available calls that it contains.
```

Author: (accepted & corrected "module module")

> To configure the runtime, open the `lib.rs` file, that contains the definition for the runtime of the included EVM template and is located in the folder:

Reviewer: Only the EVM template?

Author: Could be both included templates, reformuled the phrase to suggest that the template selection is optional

---
title: Adding a Custom-Made Module
description: Substrate is a modular blockchain framework that makes it easy to build unique and innovative Appchains composing built-in modules with custom-made ones.
---

# Adding a Custom-Made Module {: #adding-custom-made-module } 

## Introduction {: #introduction }

By providing a comprehensive library of pre-built modules addressing many common requirements, the framework greatly simplifies the process of building an Appchain and accelerates the deployment and evolution into a ContainerChain. However, addressing an innovative use case usually requires a development effort to fully meet the requirements, and, in Substrate, adding custom logic translates into writing and integrating runtime modules. 

The example presented in the [Modularity](/learn/framework/modules/#custom-module-example){target=_blank} article shows a simple lottery module exposing two transactions:

- **Buy tickets** - this function manages a user's entry into the lottery. In essence, it verifies that the participant has a sufficient balance, is not already participating, and takes care of transferring funds to register the user for the lottery

> - **Award prize**

Reviewer: ```suggestion
- **Award prize** - this function that handles a user entering into the lottery. At a high level, it fetches a pseudo random number to obtain a winner and handles the award distribution
```

---
title: Adding a Custom-Made Module
description: Substrate is a modular blockchain framework that makes it easy to build unique and innovative Appchains composing built-in modules with custom-made ones.
---

# Adding a Custom-Made Module {: #adding-custom-made-module } 

## Introduction {: #introduction }

By providing a comprehensive library of pre-built modules addressing many common requirements, the framework greatly simplifies the process of building an Appchain and accelerates the deployment and evolution into a ContainerChain. However, addressing an innovative use case usually requires a development effort to fully meet the requirements, and, in Substrate, adding custom logic translates into writing and integrating runtime modules. 

The example presented in the [Modularity](/learn/framework/modules/#custom-module-example){target=_blank} article shows a simple lottery module exposing two transactions:

> - **Buy tickets**

Reviewer: ```suggestion
- **Buy tickets** - this function manages a user's entry into the lottery. In essence, it verifies that the participant has a sufficient balance, is not already participating, and takes care of transferring funds to register the user for the lottery
```

- **Award prize**

The implementation of those transactions also uses storage, emits events, defines custom errors, and relies on other modules to handle currency (to charge for the tickets and transfer the total amount to the winner) and randomize the winner selection.

In this article, the following steps, necessary to build and add the example module to the runtime, will be covered: 

1. Create the lottery module files (package)
2. Configure the module's dependencies
3. Adding custom logic
4. Configure the runtime with the new module

--8<-- 'text/common/not-for-production-code-guard.md'

> ## Checking Prerequisites {: #checking-prerequisites }

Reviewer: In other tutorials you've not used this section, it is always good to be consistent (I like this section, but maybe add it in other tutorials)

Author: I took the idea from web3py and ethers js :)
I´ll add a ticket to add this section where missing!

To follow the steps in this guide, you will need to have the following:

> - **Clone the Tanssi repository from Github**

Reviewer: URL? Why you put it in bold?

Author: Oh, just assumed it was the default style for lists

- **Rust compiler and Cargo package manager**

> More about how to install Rust and Cargo is in the [prerequisites article](/builders/build/prerequisites/#installing-rust){target=_blank}.

Reviewer: ```suggestion
You can read more about how to install Rust and Cargo is in the [prerequisites article](/builders/build/prerequisites/#installing-rust){target=_blank}.
```

## Creating the Lottery Module Files {: #creating-lottery-module-files } 

> Before you begin writing code, you must create the files that contain the logic. Substrate modules are abstract and designed for reuse in various runtimes with different customizations. To achieve this, you'll use Cargo, which is Rust's package manager, to create the module in the form of a new package.

Reviewer: ```suggestion
Before starting your coding process, it's essential to create the files containing your logic. Substrate modules are abstract and intended for reuse across different runtimes with various customizations. To achieve this, you'll use Cargo, Rust's package manager, to create the module as a new package.
```

For instance, you can use attributes to specify details like the module's name, version, authors, and other pertinent information. For example, in the the `lottery-example` module, the `Cargo.toml` file can be configured as follows:

```toml
[package]
name = "module-lottery-example"
version = "4.0.0-dev"
description = "Simple module example"
authors = [""]
homepage = ""
...
```

This file also defines the module's dependencies, such as the core functionality that allows seamless integration with the runtime and other modules, access to storage, event emission, and more. 

The full example of the `Cargo.toml` file sets, besides the attributes, the dependencies required by Substrate:

??? code "View the complete Cargo.toml file"

    ```rust
    --8<-- 'code/basic-substrate/lottery-example-cargo.toml'
    ```

## Adding Custom Logic {: #adding-custom-logic}

As presented in the [custom-made module](/learn/framework/modules/#custom-modules){target=_blank} section of the modularity article, creating a module involves implementing the following attribute macros, of which the first three are mandatory:

> - **`#[frame_support::pallet]`** - this attribute is the entry point that marks the module as usable in the runtime

Reviewer: use snippets if this is already somewhere else

Author: I had to rewrite a little the learn article, but now that text is shared content

    TicketBought { who: T::AccountId },
    /// Event emitted when the prize is awarded
    PrizeAwarded { winner: T::AccountId },
    /// Event emitted when there are no participants
    ThereAreNoParticipants,
}
```

### Implementing Storage for State Persistence {: #implementing-storage }

The `#[pallet::storage]` macro initializes a runtime storage structure.  In the heavily constrained environment of an AppChain, deciding what to store and which structure to use can be critical in terms of performance. More on this topic is covered in the [Substrate documentation](https://docs.substrate.io/build/runtime-storage/){target=_blank}.

In this example,  the `lottery-example` module needs a basic value storage structure used to persist the list of participants in a bounded capacity vector ([BoundedVec](https://crates.parity.io/frame_support/storage/bounded_vec/struct.BoundedVec.html){target=_blank}). This can be initialized as follows:

```rust
#[pallet::storage]
#[pallet::getter(fn get_participants)]
pub(super) type Participants<T: Config> = StorageValue<
    _,
    BoundedVec<T::AccountId, T::MaxParticipants>,
    OptionQuery
>;
```
### The Complete Module {: #complete-module }

> Putting all the pieces together, after implementing all the required macros and adding the custom logic, now the module is complete and ready to be used in the runtime. 

Reviewer 2: ```suggestion
To put all the pieces together, after implementing all the required macros and adding the custom logic, the module is now complete and ready to be used in the runtime.
```

??? code "View the complete module file"

    ```rust
    --8<-- 'code/modules/lottery-example.rs'
    ```

## Configure the Runtime {: #configure-runtime }

> Finally, with the module finished, it can be included in the runtime. By doing so, the transactions `buy_tickets` and `award_prize` will be callable by the users. This also means that the (Polkadot.js API)(/builders/interact/substrate-api/polkadot-js-api/) will be decorated with this module and all the available calls that it contains.

Reviewer 2: ```suggestion
Finally, with the module finished, it can be included in the runtime. By doing so, the transactions `buy_tickets` and `award_prize` will be callable by the users. This also means that the [Polkadot.js API](/builders/interact/substrate-api/polkadot-js-api/){target=_blank} will be decorated with this module and all the available calls that it contains.
```

To configure the runtime, open the `lib.rs` file, which contains the definition for the runtime of the included template and is located (in case of using the EVM-compatible) in the folder:

```text
*/container-chains/templates/frontier/runtime/src/
```

To add the lottery module, configure the modules as follows:

```rust

// Add the configuration for randomness module. No parameters needed.
impl pallet_insecure_randomness_collective_flip::Config for Runtime {
}

// Custom module id
parameter_types! {
	pub const PalletId: PalletId = PalletId(*b"loex5678");
}

// Add configuration for the lottery module
impl pallet_lottery_example::Config for Runtime {
	type RuntimeEvent = RuntimeEvent;
	type Currency = Balances;
	type TicketCost = ConstU128<1000000000000000>;
	type PalletId = PalletId;
	type MaxParticipants = ConstU32<500>;
	type MyRandomness = RandomCollectiveFlip;
}
```

> With the modules configured, add in the macro `construct_runtime!` (that defines the modules that will be included when building the runtime) add the randomness and lottery module.

Reviewer 2: ```suggestion
With the modules configured, add the macro `construct_runtime!` (that defines the modules that will be included when building the runtime) and the randomness and lottery modules.
```

```rust
construct_runtime!(
	pub struct Runtime {
        ...
        // Include the custom logic from the pallet-template in the runtime.
        RandomCollectiveFlip: pallet_insecure_randomness_collective_flip,
        Lottery: pallet_lottery_example,
        ...
    }
)
```

> With everything set, the Apchain has now support for a basic implementation of a lottery.

Reviewer 2: ```suggestion
With everything set, the Appchain now has support for a basic implementation of a lottery.
```
